Sources/MessagingPush/PushHandling/PushClickHandler.swift

```diff
@@ -20,18 +35,29 @@ class PushClickHandlerImpl: PushClickHandler {
     }
 
     // Note: This function is called from automatic and manual push click handlers.
-    func pushClicked(_ push: PushNotification) {
-        guard let cioDelivery = push.cioDelivery else {
-            return
+    func pushClicked(_ push: PushNotification, skipPushMetricsTracking: Bool, skipDeepLinkHandling: Bool) {
```

Comment:
can't we refactor this method to not have two responsibilities? I understand its being called from both automatic, where we are already utilizing these functionalities separately. 

For manual, can we not call the two methods separately? would make codebase easier to manage and less chance of messing up stuff.

Reply:
Hopefully these [comments](https://github.com/customerio/customerio-ios/pull/590/files#diff-dfdc330b2ab4df1a2983e50045a6a6c29ce35e274532ba3eec5483b01c15056bR9-R10) could help explain the idea? 

I want to avoid this scenario: 
* In the future, we add more capabilities to when a push is clicked. Let's say for example we now track a custom event "iOS is awesome" when a push is clicked. 
* The developer goes into `iOSPushEventListener` and updates `pushAction()` function to make sure and call this new function `pushClickHandler.trackiOSIsAwesome()`
* The developer forgets to update the manual push click handling code: 
```swift
// when a push is manually clicked: 
pushClickHandling.trackOpenedMetric()
pushClickHandling.openDeepLink()
```

* I designed pushClicked to be re-usable by both automatic and manual, to avoid this scenario of forgetting to update 1 or the other. 

I would like to keep encapsulation of push clicked logic for that reason. If you have ideas for how to make the code "easier to manage and less chance of messing stuff up", could you share some ideas?

Reply:
Is this a blocking or non-blocking suggestion?

Reply:
I would say it is a blocker. We need to be careful on the design choices that can lead to mix ups and unclear responsibilities.
While this example isn't that massive, but it adds up.

Reply:
Levi, fair concern, but let's think of it this way, 

building off from your example, we have
```
pushClickHandling.trackOpenedMetric()
pushClickHandling.openDeepLink()
```

and then we have one method where you want to ensure, everything happens. 

```
func pushClicked(_ push: PushNotification) {
    // Now simply orchestrates calls without making decisions
    pushClickHandling.trackOpenedMetric()
    pushClickHandling.openDeepLink()
}
```

now we cover all the cases, if someone wants the whole `push click handling` experience vs when they want to utilize one of the APIs. 

having this doesn't remove the need to have a method that takes optional values and tries to perform multiple actions based on a condition and this API design would align more with the Single Responsibility Principle (SRP)? what do you think?

Reply:
I do agree that the API design in this PR is not ideal. With this change, the most important thing to me was to prevent breaking features by accident in the future. 

Another idea to solve this problem is leveraging automated tests. Tests that check that *all* steps for push handling is done in manual or automatic push handling. If the test breaks, it is an indicator you forgot to update something. 

If we use tests as the solution, we can then use a very simple API of just: 

```swift
pushClickHandling.trackOpenedMetric()
pushClickHandling.openDeepLink()
```

I think all 3 of us would like this simple of an API. Unless I hear differently, I'll update this PR with this solution today.

Reply:
[Made change](https://github.com/customerio/customerio-ios/compare/8133fd953cca2871897b4c7a8ea0e999248fd901..3f3fa43cca68798ae4baf726c27d654dd1fc43dd)